### PR TITLE
Reduce allocations in SslStream reading/writing by ~30%

### PR DESCRIPTION
**Pull request by stephentoub**
https://github.com/dotnet/corefx/pull/12935 reduced allocations in SslStream reading/writing by ~70%.  This PR reduces it further by another ~30% from where it was after that PR.

Before (10K reads / 10K writes):
image

After (10K reads / 10K writes):
image

Main changes:
- Every read/write operation was allocating a new AsyncProtocolRequest.  Without changing any of the plumbing through SslStream, we can take advantage of the fact that SslStream allows only a single read and a single write operation at a time, which means we can store and reuse an AsyncProtocolRequest instance for reading and one for writing, reusing the instances for all read/write operations on the stream.
- Every read operation was boxing an Int32 result.  This changes the BufferAsyncResult instance used for reads to reuse one of its existing fields to store that result, rather than storing it in the base LazyAsyncResult's object field.
- Also took the opportunity to remove an unused field from BufferAsyncResult, though it doesn't change the size on 64-bit given the other fields on the base class.

cc: @geoffkizer, @ericeil, @davidsh, @cipop, @benaadams, @davidfowl 
Contributes to https://github.com/dotnet/corefx/issues/11826

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please

File: src/System.Net.Security/src/System/Net/HelperAsyncResults.cs [hunk 1]
```diff
@@ -115,24 +132,18 @@ public bool MustCompleteSynchronously
...
 
-        internal void CompleteUser()
-        {
-            UserAsyncResult.InvokeCallback();
-        }
+        internal void CompleteUser() => UserAsyncResult.InvokeCallback();
 
-        internal void CompleteUser(object userResult)
+        internal void CompleteUser(int userResult)
         {
```
Comment (geoffkizer) [inline src/System.Net.Security/src/System/Net/HelperAsyncResults.cs]: nit: capitalization

Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/HelperAsyncResults.cs]: > nit: capitalization

Of what?

Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/HelperAsyncResults.cs]: Oh, in the debug message below?  GitHub wasn't showing that on the comment view.  Will lower-case the R.

**Comment by geoffkizer** [comment]
LGTM.  Nice work.

I do think that longer-term we should revisit the async implementation pattern here and elsewhere.  There are likely more wins to be had here.  Even with your improvements, async objects account for ~80% of allocated bytes.

**Comment by stephentoub** [comment]
> I do think that longer-term we should revisit the async implementation pattern here and elsewhere. There are likely more wins to be had here. Even with your improvements, async objects account for ~80% of allocated bytes.

Absolutely.  As you suggest, this is trying to make the best of the current design. Note that the SecBufferDesc allocation per operation can also be removed with a bit of work (see my comments in https://github.com/dotnet/corefx/pull/12935), at which point all allocations per operation will be due to asynchrony.  Once that's removed, we'll essentially have two allocations per read/write operation, one for the IAsyncResult and one for the Task wrapping it; it'll take some work, but we should be able to get that down to one, at which point I think that's generally the best we can do with the current API shape (though for some operations, we should be able to eliminate all allocations if they complete synchronously).

**Comment by benaadams** [comment]
LGTM - Excellent work! 👍

**Comment by stephentoub** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please

**Comment by stephentoub** [comment]
Thanks for the reviews, all.

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Ubuntu14.04 Debug please ("arm32 emulator vms build and test on the main drive which is small. This is tracked with issue https://github.com/dotnet/coreclr/issues/6676")

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug please (https://github.com/dotnet/corefx/issues/13281)

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug please (https://github.com/dotnet/corefx/issues/13284)

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug please (now that https://github.com/dotnet/corefx/issues/13284 is fixed)

**Pull request merged**